src/NuGet.Core/NuGet.Build.Tasks.Console/Program.cs

```diff
@@ -0,0 +1,154 @@
+using System.Threading.Tasks;
+
+namespace NuGet.Build.Tasks.Console
+{
+    /// <summary>
+    /// Represents the main entry point to the console application.
+    /// </summary>
+    internal static class Program
+    {
+        /// <summary>
+        /// A <see cref="T:char[]" /> containing the equals sign '=' to be used to split key/value pairs that are separated by it.
+        /// </summary>
+        private static readonly char[] EqualSign = { '=' };
+
+        /// <summary>
+        /// A <see cref="T:char[]" /> containing the semicolon ';' to be used to split key/value pairs that are separated by it.
+        /// </summary>
+        private static readonly char[] Semicolon = { ';' };
+
+        /// <summary>
+        /// The main entry point to the console application.
+        /// </summary>
+        /// <param name="args">The command-line arguments.</param>
+        /// <returns><code>0</code> if the application ran successfully with no errors, otherwise <code>1</code>.</returns>
+        public static async Task<int> Main(string[] args)
+        {
+            var debug = IsDebug();
+
+            if (debug)
+            {
+#if IS_CORECLR
+                System.Console.WriteLine("Waiting for debugger to attach to Process ID: {Process.GetCurrentProcess().Id}");
+
+                while (!Debugger.IsAttached)
+                {
+                    System.Threading.Thread.Sleep(100);
+                }
+
+                Debugger.Break();
+#else
+                Debugger.Launch();
+#endif
+            }
+
+            // Parse command-line arguments
+            if (!TryParseArguments(args, out var arguments))
+            {
+                return 1;
+            }
+
+            // Enable MSBuild feature flags
+            MSBuildFeatureFlags.MSBuildExeFilePath = arguments.MSBuildExeFilePath.FullName;
+            MSBuildFeatureFlags.EnableCacheFileEnumerations = true;
+            MSBuildFeatureFlags.LoadAllFilesAsReadonly = true;
+            MSBuildFeatureFlags.SkipEagerWildcardEvaluations = true;
+
+            // Only wire up an AssemblyResolve event handler if being debugged.
+            if (IsDebug())
+            {
+                // The App.config contains relative paths to MSBuild which won't work for locally built copies so an AssemblyResolve event
+                // handler is used in order to locate the MSBuild assemblies
+                string msbuildDirectory = arguments.MSBuildExeFilePath.DirectoryName;
+
+                AppDomain.CurrentDomain.AssemblyResolve += (sender, resolveArgs) =>
+                {
+                    var assemblyName = new AssemblyName(resolveArgs.Name);
+
+                    var path = Path.Combine(msbuildDirectory, $"{assemblyName.Name}.dll");
+
+                    return File.Exists(path) ? Assembly.LoadFrom(path) : null;
+                };
+            }
+
+            using (var dependencyGraphSpecGenerator = new DependencyGraphSpecGenerator(debug: debug))
```

Comment:
I think the generator should ideally just generate the dg spec, and then you can call build.task.utility to restore from here.

Reply:
`Program` can't load any MSBuild types or the assembly resolution won't work when you're debugging a locally built copy of the app.  I'll just rename my class.

Reply:
I see. Should we add a quick code comment, cause that seems like something that tests won't catch, but it could affect our debugging flow.